### PR TITLE
Adds notifications if multiple users want to edit the same motion

### LINKED ISSUE #4217: Send warning in case of simultaneously working on one motion
If more than one person is working on a motion, they should receive a warning.

### PR DESCRIPTION
**Pull request by GabrielInTheWorld**
- Created a class to define notification-objects, which should notify other persons editing the same motion.
- Added functions to send notifications, listen to them and unsubscribing them.
- Added a warning-function to the `base-view.ts`, which raises the snack bar with the given message and has no duration.

Fixes #4217

- Removed unnecessary lines of code.
- Fixed merge.

### CODE REVIEW

**Comment by GabrielInTheWorld** [comment]
Can you tell me your steps, how I can reproduce this

> Tested: Can't see any notifications. Get a TypeError in motion-detail:
> ` TypeError: Cannot read property 'open' of undefined`

Can you tell me your steps, how I could reproduce this error?

**Comment by emanuelschuetze** [comment]
> Can you tell me your steps, how I could reproduce this error?

Open the edit view of a motion  -> TypeError.

2nd Test with two admins A and B:
* A opens motion 1 and start editing.
* B (private tab) opens motion 1 too and click on "edit".
=> no notification dialog that A/B already edit the motion!

**Comment by GabrielInTheWorld** [comment]
> > Can you tell me your steps, how I could reproduce this error?
> 
> Open the edit view of a motion -> TypeError.
> 
> 2nd Test with two admins A and B:
> 
> * A opens motion 1 and start editing.
> * B (private tab) opens motion 1 too and click on "edit".
>   => no notification dialog that A/B already edit the motion!

Please test again.
Now this behaviour should be fixed.

**Comment by emanuelschuetze** [comment]
Retested your current branch:
There is still a TypeError if I go to edit motion view:
```
MotionDetailComponent.html:1 ERROR TypeError: Cannot read property 'open' of undefined
    at MotionDetailComponent.push../src/app/site/motions/components/motion-detail/motion-detail.component.ts.MotionDetailComponent.setEditMode (motion-detail.component.ts:963)
...
```

**Comment by emanuelschuetze** [comment]
I get _no_ notification dialog.

File: client/src/app/site/base/base-view.ts [hunk 1]
```diff
@@ -27,24 +27,42 @@ export abstract class BaseViewComponent extends BaseComponent implements OnDestr
...
+        if (this.matSnackBar) {
+            this.matSnackBar.dismiss();
+        }
+    }
+
     /**
      * automatically dismisses the error snack bar, if the component is destroyed.
      */
     public ngOnDestroy(): void {
-        if (this.errorSnackBar) {
```
Comment (FinnStutzenstein) [inline client/src/app/site/base/base-view.ts]: please call `claseSnackBar()`

File: client/src/app/site/motions/components/motion-detail/motion-detail.component.ts [hunk 2]
```diff
@@ -675,6 +714,8 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit {
         if (this.newMotion) {
             this.createMotion();
         } else {
+            // When saving the changes, notify other users if they edit the same motion.
+            this.unsubscribeEditNotifications(TypeOfNotificationViewMotion.TYPE_SAVING_EDITING_MOTION);
```
Comment (FinnStutzenstein) [inline client/src/app/site/motions/components/motion-detail/motion-detail.component.ts]: Please change the order: The update must be successfull to remove to publish this message.

File: client/src/app/site/motions/components/motion-detail/motion-detail.component.ts [hunk 3]
```diff
@@ -1170,6 +1218,87 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit {
...
+    /**
+     * Function to send a notification, so that other persons can recognize editing the same motion, if they're doing.
+     *
+     * @param type TypeOfNotificationViewMotion defines the type of the notification which is sent.
+     * @param user Optional userId. If set the function will send a notification to the given userId.
+     */
+    private sendEditNotification(type: TypeOfNotificationViewMotion, user?: number): void {
+        const content: ViewMotionNotificationEditMotion = {
+            motionId: this.motion.id,
+            senderId: this.randomEditorId,
```
Comment (FinnStutzenstein) [inline client/src/app/site/motions/components/motion-detail/motion-detail.component.ts]: In addition to this id, get the current username from the operator and include the short name in this notify message. The random token may not be necessary, becuase the server appends a `replayChannel` to every send message that uniqly identifies the client. Kepp a list of reply channels and user names to nicly format the message: `Users <A>, <B>, and 3 more are editing this motion`.

File: client/src/app/site/motions/components/motion-detail/motion-detail.component.ts [hunk 4]
```diff
@@ -1170,6 +1218,87 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit {
         this.repo.followRecommendation(this.motion);
     }
 
+    /**
```
Comment (FinnStutzenstein) [inline client/src/app/site/motions/components/motion-detail/motion-detail.component.ts]: Please make this more generic and move this into a service! This service just need a small interface: startEditing(<identifier, that we talk about motions>): <some kind of subscription, so the service can keep track of this> and stopEdition(<identifier>, <subscription?>, <reason>) which cleans up all subscriptions and sends the final notify message.
We maybe want to use this in other edit forms, too. And this component is already crowded enough..

File: client/src/app/site/motions/models/view-motion-notify.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/**
+ * Enum to define different types of notifications.
+ */
+export enum TypeOfNotificationViewMotion {
```
Comment (FinnStutzenstein) [inline client/src/app/site/motions/models/view-motion-notify.ts]: Move this content also to  the service and name it more generic.

File: client/src/app/site/base/base-view.ts [hunk 6]
```diff
@@ -27,24 +27,42 @@ export abstract class BaseViewComponent extends BaseComponent implements OnDestr
         super(titleService, translate);
     }
 
+    /**
+     * Opens the snack bar with the given message.
+     * This snack bar will only dismiss if the user clicks the 'OK'-button.
+     */
+    protected raiseWarning = (message: string): void => {
```
Comment (FinnStutzenstein) [inline client/src/app/site/base/base-view.ts]: Where is the difference to `raiseError`?

**Comment by FinnStutzenstein** [comment]
@emanuelschuetze THis error happens, if you try to tedit the motion while you have the mobile view. I'll create a fix for that.

File: client/src/app/site/base/base-view.ts (hunk 6 above)
Comment (GabrielInTheWorld) [reply client/src/app/site/base/base-view.ts]: This function calls a snackbar, which does not appear until the user clicks the 'okay'-button.
So, the user notices this.

File: client/src/app/site/base/base-view.ts (hunk 1 above)
Comment (GabrielInTheWorld) [reply client/src/app/site/base/base-view.ts]: Oh yes, I missed that :D

File: client/src/app/site/motions/components/motion-detail/motion-detail.component.ts (hunk 2 above)
Comment (tsiegleauq) [reply client/src/app/site/motions/components/motion-detail/motion-detail.component.ts]: calling `updateMotionFromForm` should be sufficient I suppose

File: client/src/app/site/motions/components/motion-detail/motion-detail.component.ts (hunk 3 above)
Comment (tsiegleauq) [reply client/src/app/site/motions/components/motion-detail/motion-detail.component.ts]: done

**Comment by tsiegleauq** [comment]
- cleaned up some code
- transfer short names over notify
- show the names of currently editing users

@FinnStutzenstein I would advise to create the required service in another PR
please review again

**Comment by tsiegleauq** [comment]
also @emanuelschuetze

**Pull request merged**